packages/shared_preferences/shared_preferences/lib/shared_preferences.dart

```diff
@@ -14,12 +14,39 @@ import 'package:shared_preferences_platform_interface/shared_preferences_platfor
 class SharedPreferences {
   SharedPreferences._(this._preferenceCache);
 
-  static const String _prefix = 'flutter.';
+  static String _currentPrefix = 'flutter.';
+
+  static bool _prefixHasBeenSet = false;
+
   static Completer<SharedPreferences>? _completer;
 
   static SharedPreferencesStorePlatform get _store =>
       SharedPreferencesStorePlatform.instance;
 
+  /// Sets the prefix that is attached to all keys for all shared preferences.
+  ///
+  /// This changes the inputs when adding data to preferences as well as
+  /// setting the filter that determines what data will be returned
+  /// from the `getInstance` method.
+  ///
+  /// This cannot be called after `getInstance`.
```

Comment:
Why not modify `getInstance` so it's:

`Future<plugin> getInstance([String? prefix])`

?

That way the plugin API wouldn't need a `getAllWithPrefix()` (maybe a `getAll([String? prefix])`?

Reply:
> Why not modify `getInstance` so it's:
> 
> `Future<plugin> getInstance([String? prefix])`

The issue is that `shared_prferences` is designed as a singleton, with `getInstance` getting the single instance. So:

```
// Get one instance with the legacy prefix.
var instance1 = await SharedPreferences.getInstance();
// Get another instance without prefixes.
var instance2 = await SharedPreferences.getInstance('');
// Now migrate prefs between them!
[...]
```

Would not do at all what it looks like it would do, which would be super confusing. And if we change it to not be a singleton to fix that, we open up a whole can of worms around the fact that most of it operates on a Dart-side cache, not the actual underlying store, so you'd have consistency problems.

Longer term we may need to revisit the public API in a more significant way (e.g., maybe we provide a public fully async API, which does allow instantiation with different prefixes, and the cache is an optional layer you can add on to an instance and manage yourself), but this was the least bad way we could find to add this incrementally.

Reply:
Ah man, I was confused by the static factory method, I thought it returned a new instance of the plugin object (like in your second example!)

We need a "The Future Of The shared_preferences Plugin" issue, it seems :P

Reply:
> We need a "The Future Of The shared_preferences Plugin" issue, it seems :P

Yep, I filed https://github.com/flutter/flutter/issues/123078